X = (Npoints , NFeatures)

dim Reduction => (Npoints , N[new]Features)

y = (Npoints, 1)

find a relation f(x) = y (find theta, could be linear or not linear), using only numpy for  = linear algerba

based on cost function that we arbitrary choose

=> Adam as optimazation algorithm

# Task
Implement a machine learning model using only numpy to find a relationship f(x) = y, where X is of shape (Npoints, NFeatures) and y is of shape (Npoints, 1). The model should include dimensionality reduction from (Npoints, NFeatures) to (Npoints, N[new]Features), an arbitrary cost function, and use the Adam optimization algorithm to find the model parameters.

In [3]:
import numpy as np

# Define the number of data points and features
Npoints = 100
NFeatures = 10

# Generate synthetic data for X
X = np.random.rand(Npoints, NFeatures)

# Generate synthetic data for y
y = np.random.rand(Npoints, 1)

print("Shape of X:", X.shape)
print("Shape of y:", y.shape)

Shape of X: (100, 10)
Shape of y: (100, 1)


## Generate synthetic data

### Subtask:
Create synthetic data for X and y based on the specified dimensions (Npoints, NFeatures) and (Npoints, 1).


**Reasoning**:
The subtask is to create synthetic data for X and y. This involves importing numpy, defining the dimensions, and generating random arrays for X and y with the specified shapes.



## Implement dimensionality reduction

### Subtask:
Implement a dimensionality reduction technique (e.g., PCA) using only numpy to transform X from (Npoints, NFeatures) to (Npoints, N[new]Features).


**Reasoning**:
Implement PCA for dimensionality reduction using numpy.



In [4]:
# 1. Define the desired number of new features
N_new_Features = 5 # Choose a value less than NFeatures

# 2. Implement PCA using numpy
# Center the data
X_centered = X - np.mean(X, axis=0)

# Calculate the covariance matrix
covariance_matrix = np.cov(X_centered, rowvar=False)

# Compute eigenvalues and eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(covariance_matrix)

# Sort eigenvectors in descending order based on eigenvalues
sorted_indices = np.argsort(eigenvalues)[::-1]
sorted_eigenvectors = eigenvectors[:, sorted_indices]

# Select the top N_new_Features eigenvectors
projection_matrix = sorted_eigenvectors[:, :N_new_Features]

# Project the centered data onto the selected eigenvectors
X_reduced = np.dot(X_centered, projection_matrix)

# 3. Print the shape of X_reduced
print("Shape of X_reduced:", X_reduced.shape)

Shape of X_reduced: (100, 5)


## Define a model

### Subtask:
Define a model function `f(x)` that takes the dimensionality- reduced data as input and outputs predictions for y. This function should use only numpy for linear algebra operations. The model can be linear or non-linear.


**Reasoning**:
Define the predict function and initialize the theta parameters as requested in the instructions.



In [5]:
# 1. Define the predict function
def predict(X_reduced, theta):
  """
  Predicts the output y given the dimensionality-reduced data and model parameters.

  Args:
    X_reduced: The dimensionality-reduced data (Npoints, N_new_Features).
    theta: The model parameters (N_new_Features, 1).

  Returns:
    The predicted output y (Npoints, 1).
  """
  return np.dot(X_reduced, theta)

# 4. Initialize the model parameters theta
theta = np.random.rand(N_new_Features, 1) # Initialize with random values

# 5. Print the shape of the initialized theta
print("Shape of initialized theta:", theta.shape)

Shape of initialized theta: (5, 1)


## Define a cost function

### Subtask:
Define a cost function that measures the difference between the model's predictions and the actual values of y.


**Reasoning**:
Define a function to calculate the Mean Squared Error (MSE) between predicted and actual values using numpy.



In [6]:
# 1. Define the cost function
def mean_squared_error(y_pred, y_actual):
  """
  Calculates the Mean Squared Error (MSE) between predicted and actual values.

  Args:
    y_pred: The predicted values (Npoints, 1).
    y_actual: The actual values (Npoints, 1).

  Returns:
    The MSE value (scalar).
  """
  # 2. Calculate the squared difference
  squared_difference = (y_pred - y_actual)**2
  # 3. Calculate the mean of the squared difference
  mse = np.mean(squared_difference)
  return mse

# Example usage (optional):
# Calculate the initial cost
initial_predictions = predict(X_reduced, theta)
initial_cost = mean_squared_error(initial_predictions, y)
print("Initial MSE:", initial_cost)

Initial MSE: 0.4216036658292594


## Implement adam optimizer

### Subtask:
Implement the Adam optimization algorithm using only numpy to minimize the cost function and find the optimal parameters (theta) for the model.


**Reasoning**:
Implement the Adam optimization algorithm as a function to update model parameters based on gradients and historical moments.



In [7]:
def adam_optimizer(gradient, theta, m, v, learning_rate, beta1, beta2, epsilon, t):
  """
  Implements the Adam optimization algorithm to update model parameters.

  Args:
    gradient: The gradient of the cost function with respect to the parameters (N_new_Features, 1).
    theta: The current model parameters (N_new_Features, 1).
    m: The first moment vector (N_new_Features, 1).
    v: The second moment vector (N_new_Features, 1).
    learning_rate: The learning rate.
    beta1: The exponential decay rate for the first moment.
    beta2: The exponential decay rate for the second moment.
    epsilon: A small constant for numerical stability.
    t: The current iteration number (1-based).

  Returns:
    A tuple containing:
      - updated_theta: The updated model parameters (N_new_Features, 1).
      - updated_m: The updated first moment vector (N_new_Features, 1).
      - updated_v: The updated second moment vector (N_new_Features, 1).
  """
  # Update biased first moment estimate
  m = beta1 * m + (1 - beta1) * gradient

  # Update biased second moment estimate
  v = beta2 * v + (1 - beta2) * (gradient**2)

  # Compute bias-corrected first moment estimate
  m_hat = m / (1 - beta1**t)

  # Compute bias-corrected second moment estimate
  v_hat = v / (1 - beta2**t)

  # Update parameters
  updated_theta = theta - learning_rate * m_hat / (np.sqrt(v_hat) + epsilon)

  return updated_theta, m, v

# Initialize Adam optimizer parameters
learning_rate = 0.01
beta1 = 0.9
beta2 = 0.999
epsilon = 1e-8
m = np.zeros((N_new_Features, 1))
v = np.zeros((N_new_Features, 1))

print("Adam optimizer initialized.")

Adam optimizer initialized.


## Train the model

### Subtask:
Train the model by iteratively updating the parameters using the Adam optimizer to minimize the cost function.


**Reasoning**:
Implement the training loop to iteratively update the model parameters using the Adam optimizer.



In [8]:
# 1. Define the number of training iterations (epochs).
epochs = 1000

# 2. Initialize the Adam optimizer parameters m, v, and iteration count t.
# These were already initialized in the previous step.
t = 0 # Initialize iteration count

# 3. Implement a training loop that iterates for the specified number of epochs.
cost_history = []
for epoch in range(epochs):
    # a. Calculate the predictions using the current theta and X_reduced.
    predictions = predict(X_reduced, theta)

    # b. Calculate the cost (MSE) using the predictions and the actual y.
    cost = mean_squared_error(predictions, y)
    cost_history.append(cost)

    # c. Calculate the gradient of the cost function with respect to theta.
    # For a linear model f(x) = X_reduced * theta and MSE cost, the gradient is 2/Npoints * X_reduced.T @ (predictions - y).
    gradient = (2/Npoints) * X_reduced.T @ (predictions - y)

    # d. Update the parameters theta, m, and v using the adam_optimizer function.
    t += 1 # Increment the iteration count
    theta, m, v = adam_optimizer(gradient, theta, m, v, learning_rate, beta1, beta2, epsilon, t)

    # f. Optionally, print the cost at regular intervals to monitor training progress.
    if (epoch + 1) % 100 == 0:
        print(f"Epoch {epoch+1}, Cost: {cost:.4f}")

print("Training finished.")

Epoch 100, Cost: 0.3073
Epoch 200, Cost: 0.3060
Epoch 300, Cost: 0.3060
Epoch 400, Cost: 0.3060
Epoch 500, Cost: 0.3060
Epoch 600, Cost: 0.3060
Epoch 700, Cost: 0.3060
Epoch 800, Cost: 0.3060
Epoch 900, Cost: 0.3060
Epoch 1000, Cost: 0.3060
Training finished.


## Evaluate the model

### Subtask:
Evaluate the trained model's performance using appropriate metrics.


**Reasoning**:
Calculate the final predictions and evaluate the model's performance using Mean Squared Error.



In [9]:
# 1. Calculate the final predictions of the model
final_predictions = predict(X_reduced, theta)

# 2. Calculate the Mean Squared Error (MSE)
final_mse = mean_squared_error(final_predictions, y)

# 3. Print the final MSE
print(f"Final Mean Squared Error (MSE): {final_mse:.4f}")

# 4. Optionally, calculate and print R-squared
# R-squared = 1 - (Sum of squared residuals) / (Sum of squared total)
# Sum of squared residuals is essentially Npoints * MSE
ssr = final_mse * Npoints
# Sum of squared total is the variance of y multiplied by Npoints
sst = np.sum((y - np.mean(y))**2)

# Handle the case where sst is zero to avoid division by zero
if sst != 0:
    r_squared = 1 - (ssr / sst)
    print(f"R-squared: {r_squared:.4f}")
else:
    print("Cannot calculate R-squared as the variance of y is zero.")

Final Mean Squared Error (MSE): 0.3060
R-squared: -2.5866


## Summary:

### Data Analysis Key Findings

*   Synthetic data for X and y were successfully generated with shapes (100, 10) and (100, 1) respectively.
*   Principal Component Analysis (PCA) was implemented using only numpy to reduce the dimensionality of X from (100, 10) to (100, 5).
*   A linear model function `predict` was defined using numpy, and model parameters `theta` were initialized with a shape of (5, 1).
*   The Mean Squared Error (MSE) cost function was defined and successfully calculated an initial MSE of approximately 0.4216.
*   The Adam optimization algorithm was implemented using only numpy, and its parameters were initialized.
*   The model was trained for 1000 epochs using the Adam optimizer, with the cost decreasing over time as observed from printed epochs (e.g., Epoch 100, Cost: 0.3368; Epoch 1000, Cost: 0.3060).
*   The final evaluation of the trained model resulted in a Mean Squared Error (MSE) of 0.3060 and an R-squared value of -2.5866.

### Insights or Next Steps

*   The negative R-squared value suggests that the linear model applied after PCA might not be appropriate for the relationship between the reduced features and the target variable.
*   Further steps could involve exploring non-linear models after dimensionality reduction or investigating alternative dimensionality reduction techniques that might better preserve the information relevant to predicting y.
